pkg/catalog/manager/traverse.go

```diff
@@ -4,23 +4,41 @@ import (
 	"fmt"
 	"strings"
 
+	"github.com/pkg/errors"
 	"github.com/rancher/rancher/pkg/catalog/helm"
 	"github.com/rancher/types/apis/management.cattle.io/v3"
 	"github.com/sirupsen/logrus"
 	"gopkg.in/yaml.v2"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
 )
 
-func traverseFiles(repoPath string, catalog *v3.Catalog) ([]v3.Template, []string, map[string]v3.VersionCommits, []error, error) {
+func (m *Manager) traverseAndUpdate(repoPath, commit string, catalog *v3.Catalog) error {
 	logrus.Info("traversing files and generating templates")
 	index, err := helm.LoadIndex(repoPath)
 	if err != nil {
-		return nil, nil, nil, nil, err
+		return err
+	}
+
+	// list all existing templates
+	templateMap, err := m.getTemplateMap(catalog.Name)
+	if err != nil {
+		return err
+	}
+	// list all templateContent tag
+	templateContentMap := map[string]struct{}{}
+	templateContentList, err := m.templateContentLister.List("", labels.NewSelector())
+	if err != nil {
+		return err
+	}
+	for _, t := range templateContentList {
+		templateContentMap[t.Name] = struct{}{}
 	}
-	newHelmVersionCommits := map[string]v3.VersionCommits{}
 
-	var templates []v3.Template
-	var errors []error
+	newHelmVersionCommits := map[string]v3.VersionCommits{}
+	var errs []error
+	var terrors []error
 	for chart, metadata := range index.IndexFile.Entries {
 		newHelmVersionCommits[chart] = v3.VersionCommits{
```

Comment:
Any reason why we decided on sequential load? Have we considering using errgroup, and run download in separate go routines? Something like [here] (https://github.com/rancher/rke/blob/8b8bc8ab8beb2b38546474223bbd479debd7082b/cluster/logs.go#L19)